### PR TITLE
hcoll: derived datatype support

### PR DESCRIPTION
**Pull request by janjust**

### CODE REVIEW

**Comment by raffenet** [comment]
test:jenkins/ch4/ucx

File: src/mpid/ch4/src/ch4_comm.h [hunk 1]
```diff
@@ -201,6 +203,10 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_Comm_create_hook(MPIR_Comm * comm)
         }
     }
 
+#if defined HAVE_LIBHCOLL
+    hcoll_comm_create(comm, NULL);
+#endif
+
```
Comment (wesbland) [inline src/mpid/ch4/src/ch4_comm.h]: Given that hcoll is UCX-specific for now, should this be in `ucx_comm.h`?

Comment (janjust) [inline src/mpid/ch4/src/ch4_comm.h]: It'll require some more changes in order to move it to ucx layer - for now we're just trying to validate the PR

**Comment by raffenet** [comment]
test:jenkins/ch4/ucx

File: src/mpid/ch4/src/ch4_comm.h (hunk 1 above)
Comment (raffenet) [inline src/mpid/ch4/src/ch4_comm.h]: I think there are some valid reasons for doing the rank-map detection/creation prior to calling the NM/SHM hooks here. @yfguo is looking into if this is OK. I think @ddurnov is also interested in this change.

Comment (raffenet) [inline src/mpid/ch4/src/ch4_comm.h]: In which case, I am assuming we can safely move this `hcoll_comm_create` down into the UCX NM hook.

**Comment by raffenet** [comment]
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
test:jenkins/ch3/mxm

**Comment by raffenet** [comment]
@janjust @vspetrov I'm able to build this now for both ch3/ch4. A simple test shows this error, though: ```[:23238][hcoll_collectives.c:380:hcoll_create_context] Error: runtime has not provided an appropriate tag offset for the ptp runtime APIs
HCOLL context can not be created: Returning NULL.
```
Any suggestions?

**Comment by raffenet** [comment]
Oh, dumb mistake. I had forgetten to set `HCOLL_ENABLE=1`.

File: src/mpid/common/hcoll/hcoll_dtypes.h [hunk 2]
```diff
@@ -44,6 +57,124 @@ static dte_data_representation_t mpi_dtype_2_dte_dtype(MPI_Datatype datatype)
...
+        MPIDU_Datatype_get_size_macro(datatype, size);
+        switch(size) {
+            case 4:
+                return DTE_2FLOAT32;
+            case 8:
+                return DTE_2FLOAT64;
+            default:
+                return DTE_ZERO;
+        }
+    case MPI_2DBLPREC:
```
Comment (raffenet) [inline src/mpid/common/hcoll/hcoll_dtypes.h]: should be MPI_2DOUBLE_PRECISION

**Comment by raffenet** [comment]
@janjust I've rebased your dtype changes on the current mpich/master branch. Is it OK if I force push here? That way this PR will change topic to non-contig datatype support with HCOLL.

**Comment by pavanbalaji** [comment]
Is this superseded by #2740?

**Comment by janjust** [comment]
@raffenet @pavanbalaji This PR should be rebased on top of #2740.
@raffenet I think we decided to wait until after the collective refactor before merging the dtype support, correct?

**Comment by raffenet** [comment]
Correct. The coll infrastructure stuff is done, so we should get this back in shape. I seem to have done a rebase at some point in the past but never pushed it back up to github. I'll dig around on my machine and see if there's anything worth salvaging.

**Comment by raffenet** [comment]
Rebased and cleaned up a bit. Still need to logically separate some of these commits, but wanted to get a testing baseline.
test:jenkins/ch3/mxm
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
Added and squashed some fixups. Should now properly ignore send type when `MPI_IN_PLACE` is used.
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
test:jenkins/ch4/ucx

**Comment by raffenet** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/most

**Comment by raffenet** [comment]
Some of these tests failed because of a bug in the test script. Re-running...
test:jenkins/ch3/tcp
test:jenkins/ch4/most

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/most

**Comment by raffenet** [comment]
test:jenkins/ch4/most

**Pull request merged**